htdocs/product/class/product.class.php

```diff
@@ -5830,6 +5830,22 @@ public function load_stock($option = '', $includedraftpoforvirtual = null, $date
 			}
 		}
 
+		if (! preg_match('/novirtual/', $option)) {
+			// initialize $this->stock_warehouse for all warehouses, even those with no current physical stock because
+			// they may still have a non-zero virtual stock
+			$sql = "SELECT rowid FROM " . $this->db->prefix() . "entrepot WHERE entity IN (" . getEntity('stock') . ")";
+			$result = $this->db->query($sql);
+			if ($result) {
+				while ($obj = $this->db->fetch_object($result)) {
+					$stockWarehouse = $this->stock_warehouse[$obj->rowid] = new stdClass();
+					$stockWarehouse->id = 0;
+					$stockWarehouse->real = 0;
+					$stockWarehouse->virtual = 0;
+					$stockWarehouse->detail_batch = array();
+				}
+			}
+		}
+
```

Comment:
Actually, for the second problem, it would be better to do this initialization regardless of whether virtual stock is requested or not.

```suggestion
	// initialize $this->stock_warehouse for all warehouses, even those with no current physical stock because
	// they may still have a non-zero virtual stock
	$sql = "SELECT rowid FROM " . $this->db->prefix() . "entrepot WHERE entity IN (" . getEntity('stock') . ")";
	$result = $this->db->query($sql);
	if ($result) {
		while ($obj = $this->db->fetch_object($result)) {
			$stockWarehouse = $this->stock_warehouse[$obj->rowid] = new stdClass();
			$stockWarehouse->id = 0;
			$stockWarehouse->real = 0;
			$stockWarehouse->virtual = 0;
			$stockWarehouse->detail_batch = array();
		}
	}
```